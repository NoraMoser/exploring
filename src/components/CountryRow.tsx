import React from "react";
import { Country } from "../types/Country";

type CountryRowProps = {
  country: Country;
  //singld country object from the array of countries
  onSelectCountry: (code: string) => void;
  //function to set the specific country by the country code used with the handlekeydown function so the user can use the keyboard to select
};

const CountryRow: React.FC<CountryRowProps> = ({
  country,
  onSelectCountry,
}) => {
  const handleKeyDown = (e: React.KeyboardEvent) => {
    //want to make sure you can do everything by keyboard and hitting enter and not just mouse for accessibility
    if (e.key === "Enter") {
      onSelectCountry(country.cca3);
    }
  };

  return (
    <tr
      tabIndex={0}
      onClick={() => onSelectCountry(country.cca3)}
      onKeyDown={handleKeyDown}
      style={{ cursor: "pointer" }}
      aria-label={`view details for ${country.name.common}`} //for screen readers ti be able to access
      data-testid={`row-${country.cca3}`} //help with testing
    >
      <td data-label="Common Name" title={country.name.common}>
        {country.name.common}
      </td>
      <td data-label="Official Name" title={country.name.official}>
        {country.name.official ?? "N/A"}
        {/* in case some countries do not have an official name */}
      </td>
    </tr>
  );
};

export default CountryRow;
