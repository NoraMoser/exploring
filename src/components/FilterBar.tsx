import React from 'react'


type FilterProps = {
  setQuery: (query: string) => void;
  query: string
};

const FilterBar: React.FC<FilterProps> = ({setQuery, query}) => {

    return (
        <nav>
            <input
              className="search"
              type="text"
              placeholder="Search countries..."
              value={query}
              onChange={(e) => setQuery(e.target.value)}
            />
        </nav>
    );
}

export default FilterBar