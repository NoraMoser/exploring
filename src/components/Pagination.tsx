import React from "react";
import { Country } from "../types/Country";

type PaginationProps = {
  currentPage: number;
  //this is the current page the the user is looking at on the home screen
  setCurrentPage: (currentPage: number) => void;
  //this is so the user can change the currentpage state when the button is clicked
  amtCountriesPerPage: number;
  //this is a value that tells how many items in the list the user can see per page
  filteredCountries: Country[];
  //this shows the countries list that has been filtered by the search bar and also ordered alphabetically
};

const Pagination: React.FC<PaginationProps> = ({
  currentPage,
  setCurrentPage,
  amtCountriesPerPage,
  filteredCountries,
}) => {
  return (
    <div className="pagination">
      <button
        onClick={() => setCurrentPage(currentPage - 1)}
        disabled={currentPage === 1}
      >
        Previous
      </button>
      <span>
        {/* math to tell how many total pages there are */}
        Page {currentPage} of{" "} {/* prettier keeps doing it like this */}
        {Math.ceil(filteredCountries.length / amtCountriesPerPage)}
      </span>
      <button
        onClick={() => setCurrentPage(currentPage + 1)}
        disabled={
          currentPage ===
          Math.ceil(filteredCountries.length / amtCountriesPerPage)
        }
      >
        Next
      </button>
    </div>
  );
};

export default Pagination;
