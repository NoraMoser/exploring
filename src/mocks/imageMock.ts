
const mockImage = 'test-file-stub.png';
export default mockImage;
//this is to fake the marker icon for leaflet for the test 