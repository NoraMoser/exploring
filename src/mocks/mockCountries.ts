// mockCountries.ts
import { Country } from "../types/Country"; // adjust the path if needed

export const mockCountries: Country[] = [
  {
    name: {
      common: "Canada",
      official: "Canadadada",
    },
    cca3: "CAN",
    capital: ["Ottawa"],
    region: "Americas",
    subregion: "Northern America",
    population: 38000000,
    flags: {
      svg: "https://flagcdn.com/ca.svg",
      alt: "The flag of Canada features a red maple leaf in the center with red borders on either side.",
    },
    maps: {
      googleMaps: "https://goo.gl/maps/jGnPRdZ5rfsCkS2E6",
      openStreetMaps: "https://www.openstreetmap.org/relation/1428125",
    },
    languages: {
      eng: "English",
      fra: "French",
    },
    latlng: [56.1304, -106.3468],
    coatOfArms: {
      svg: "https://mainfacts.com/media/images/coats_of_arms/ca.svg",
    },
    area: 9984670,
    currencies: {
      CAD: {
        name: "Canadian dollar",
        symbol: "$",
      },
    },
    timezones: ["UTC−08:00", "UTC−07:00", "UTC−06:00", "UTC−05:00", "UTC−04:00", "UTC−03:30"],
    continents: ["North America"],
  },
  {
    name: {
      common: "France",
      official: "French Republic",
    },
    cca3: "FRA",
    capital: ["Paris"],
    region: "Europe",
    subregion: "Western Europe",
    population: 67000000,
    flags: {
      svg: "https://flagcdn.com/fr.svg",
      alt: "The flag of France is composed of three vertical bands colored blue, white, and red.",
    },
    maps: {
      googleMaps: "https://goo.gl/maps/g7QxxSFsWyTPKuzd7",
      openStreetMaps: "https://www.openstreetmap.org/relation/1403916",
    },
    languages: {
      fra: "French",
    },
    latlng: [46.6034, 1.8883],
    coatOfArms: {
      svg: "https://mainfacts.com/media/images/coats_of_arms/fr.svg",
    },
    area: 551695,
    currencies: {
      EUR: {
        name: "Euro",
        symbol: "€",
      },
    },
    timezones: ["UTC+01:00"],
    continents: ["Europe"],
  },
];
