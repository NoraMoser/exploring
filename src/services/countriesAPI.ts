import { Country } from "../types/Country";

const BASE_URL = "https://restcountries.com/v3.1";
//you can find these endpoints at https://restcountries.com/#endpoints

export const fetchAllCountries = async (): Promise<Country[]> => {
  const response = await fetch(`${BASE_URL}/all`);
  if (!response.ok) {
    throw new Error("Unable to fetch countries");
  }
  return await response.json();
};

export const fetchCountryByCode = async (code: string): Promise<Country> => {
  const response = await fetch(`${BASE_URL}/alpha/${code}`);
  if (!response.ok) {
    throw new Error("Unable to fetch country");
  }
  const data = await response.json();
  return data[0];
};
