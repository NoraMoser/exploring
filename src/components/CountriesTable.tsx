import React from "react";
import { Country } from "../types/Country";
import CountryRow from "./CountryRow";
import { useNavigate } from "react-router-dom";

type CountriesTableProps = {
  filteredCountries: Country[];
  // Filtered list of countries used in the table
  currentPage: number;
  //the current page the user is on (changes when they push the prev or next buttons)
  amtCountriesPerPage: number;
  //amount of items in the list that are displayed at one time
};

const CountriesTable: React.FC<CountriesTableProps> = ({
  filteredCountries,
  currentPage,
  amtCountriesPerPage,
}) => {
  const navigate = useNavigate();

  const indexOfFirstCountry = currentPage * amtCountriesPerPage;
  const indexOfLastCountry = indexOfFirstCountry - amtCountriesPerPage;
  //these are for the pagination so that we know where to slice- cannot slice immediately bc it messes up the pagination total pages so just do it before we list

  const handleSelectCountry = (code: string) => {
    navigate(`/country/${code}`);
    //passing this country code in the url which i have defined in the react router - in app.tsx
  };

  return (
    <table role="presentation"  className="country-table">
      <thead>
        <tr>
          <th>Common Name</th>
          <th>Official Name</th>
        </tr>
      </thead>
      <tbody>
        {filteredCountries.length > 0 ? (
          filteredCountries
            .slice(indexOfLastCountry, indexOfFirstCountry)
            .map((country) => (
              <CountryRow
                key={country.cca3}
                country={country}
                onSelectCountry={handleSelectCountry}
              />
            ))
        ) : (
          <tr>
            <td colSpan={2} style={{ textAlign: "center", padding: "1rem" }}>
              Sorry, no countries found.
            </td>
          </tr>
          // In case the user searches for a country that is not there
        )}
      </tbody>
    </table>
  );
};
//this is the list of all the countries displayed in a table
export default CountriesTable;
