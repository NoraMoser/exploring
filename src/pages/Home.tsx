import React, { useEffect, useMemo, useState } from "react";
import { useQuery } from "@tanstack/react-query";
import { fetchAllCountries } from "../services/countriesAPI";

//components
import FilterBar from "../components/FilterBar";
import Pagination from "../components/Pagination";
import CountriesTable from "../components/CountriesTable";

const Home = () => {
  const {
    data: countries,
    isLoading,
    error,
  } = useQuery({
    queryKey: ["countries"],
    queryFn: fetchAllCountries,
  });
  //using react query here to fetch the data from the function in countriesapi
  const [query, setQuery] = useState(""); //this is the value from the search input
  const [currentPage, setCurrentPage] = useState(1);

  const amtCountriesPerPage = 10;
  const filteredCountries = useMemo(() => {
    //this caches filtered countries so it will run only with a change of one of the vars in the dependency array
    if (!countries) return [];

    const sortedCountries = countries.sort((a, b) =>
      a.name.common.localeCompare(b.name.common)
    ); //this puts names in alphabetical order

    const filtered = query
      ? sortedCountries.filter(
          (country) =>
            country.name.common
              .toLowerCase()
              .includes(query.trim().toLowerCase()) //no white space or upper case letters
        )
      : sortedCountries; //if the query is more than an empty string, we filter based on the countries that contain that letter

    return filtered;
  }, [countries, query]);

  useEffect(() => {
    setCurrentPage(1);
  }, [query]); //if you are not on page one and try to search, it need to go back to page 1 or it won't work bc you might be on the wrong page to see the results

  if (isLoading || !countries) return <p>Loading...</p>;
  if (error) return <p>Error loading countries.</p>;

  const handleSearchInput = (name: string) => {
    setQuery(name);
  };

  return (
    <div>
      <FilterBar query={query} setQuery={handleSearchInput} />
      <CountriesTable
        filteredCountries={filteredCountries}
        currentPage={currentPage}
        amtCountriesPerPage={amtCountriesPerPage}
      />
      <Pagination
        amtCountriesPerPage={amtCountriesPerPage}
        currentPage={currentPage}
        setCurrentPage={setCurrentPage}
        filteredCountries={filteredCountries}
      />
    </div>
  );
};

export default Home;
