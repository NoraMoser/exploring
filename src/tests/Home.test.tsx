import React from "react";
import { render, screen, waitFor, fireEvent } from "@testing-library/react";
import { BrowserRouter, MemoryRouter, useNavigate } from "react-router-dom";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { mockCountries } from "../mocks/mockCountries";
import { fetchAllCountries } from "../services/countriesAPI";
import FilterBar from "../components/FilterBar";
import Home from "../pages/Home";
import CountriesTable from "../components/CountriesTable";
import CountryRow from "../components/CountryRow";
import Pagination from "../components/Pagination";

const mockedNavigate = jest.fn();

jest.mock("react-router-dom", () => {
    const actual = jest.requireActual("react-router-dom");
    return {
      ...actual,
      useNavigate: () => mockedNavigate,
    };
  });
  

jest.mock("../services/countriesAPI", () => ({ //mock the countries api
  fetchAllCountries: jest.fn(),
}));

beforeEach(() => {
  (fetchAllCountries as jest.Mock).mockResolvedValue(mockCountries);
});

const createTestQueryClient = () =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });
const queryClient = createTestQueryClient();

const renderWithRouter = (ui: React.ReactElement) =>
  render(
    <BrowserRouter>
      <QueryClientProvider client={queryClient}>{ui}</QueryClientProvider>
    </BrowserRouter>
  );

//Home Page**********
test("renders loading state initially", () => {
  //before getting the api, makes sure the loading page is showing
  renderWithRouter(<Home />);
  expect(screen.getByText(/loading/i)).toBeInTheDocument();
});

test("renders list of countries after fetch", async () => {
  //wait for the api to load an then check for a name
  renderWithRouter(<Home />);
  // Wait for some country name from mockCountries to appear
  await waitFor(() => {
    const cells = screen.getAllByText("Canada"); //each common name is in both title and the written text
    expect(cells.length).toBeGreaterThan(0);
  });
});

test("renders FilterBar component inside Home", () => {
  renderWithRouter(<Home />);
  // For example, the FilterBar's heading "exploring"
  const filterBarHeading = screen.getByRole("heading", { name: /exploring/i });
  expect(filterBarHeading).toBeInTheDocument();
});

//Filter Bar Component******
test("renders heading from FilterBar with fake props", () => {
  //want to make sure the main title is there
  const fakeProps = {
    query: "Nora",
    setQuery: () => {},
  };
  render(<FilterBar {...fakeProps} />);
  const heading = screen.getByRole("heading", { name: /exploring/i });
  expect(heading).toBeInTheDocument();
});

test("renders input with placeholder 'search countries...' in FilterBar", () => {
  const fakeProps = {
    query: "Nora",
    setQuery: () => {},
  };
  render(<FilterBar {...fakeProps} />);
  const input = screen.getByPlaceholderText(/search countries.../i);
  expect(input).toBeInTheDocument();
});

test("calls setQuery when input changes", () => {
  //tests that the function is called when the user types in the input bar
  const setQuery = jest.fn();
  render(<FilterBar query="" setQuery={setQuery} />);

  const input = screen.getByPlaceholderText(/search countries.../i);
  fireEvent.change(input, { target: { value: "Nora" } });

  expect(setQuery).toHaveBeenCalledWith("Nora");
});

//CountriesTable Component**********

test("renders a table with headings", () => {
  const fakeProps = {
    filteredCountries: mockCountries,
    currentPage: 1,
    amtCountriesPerPage: 10,
  };
  render(
    <MemoryRouter>
      <CountriesTable {...fakeProps} />
    </MemoryRouter>
  ); //you have to wrap in memory router due to useNavigate on this component
  expect(
    screen.getByRole("columnheader", { name: /common name/i })
  ).toBeInTheDocument();
});

test("shows the correct country name after slicing", () => {
  const fakeProps = {
    filteredCountries: mockCountries,
    currentPage: 1,
    amtCountriesPerPage: 1,
  };
  render(
    <MemoryRouter>
      <CountriesTable {...fakeProps} />
    </MemoryRouter>
  ); //you have to wrap in memory router due to useNavigate on this component
  const cells = screen.getAllByText("Canada"); //each common name is in both title and the written text
  expect(cells.length).toBeGreaterThan(0);
  expect(screen.queryByText("France")).not.toBeInTheDocument();
});

test("shows no countries message when list is empty", () => {
  const fakeProps = {
    filteredCountries: [],
    currentPage: 1,
    amtCountriesPerPage: 10,
  };
  render(
    <MemoryRouter>
      <CountriesTable {...fakeProps} />
    </MemoryRouter>
  );
  expect(screen.getByText(/sorry, no countries found/i)).toBeInTheDocument();
});

test("if list is not empty, do not show no countries message when list is empty", () => {
  const fakeProps = {
    filteredCountries: mockCountries,
    currentPage: 1,
    amtCountriesPerPage: 10,
  };
  render(
    <MemoryRouter>
      <CountriesTable {...fakeProps} />
    </MemoryRouter>
  );
  expect(screen.queryByText(/sorry, no countries found/i)).not.toBeInTheDocument();
});

test("navigates to country detail on row click", () => { //mocked useNavigate above so just making sure it is working
    const fakeProps = {
      filteredCountries: mockCountries,
      currentPage: 1,
      amtCountriesPerPage: 10,
    };
    render(
      <MemoryRouter>
        <CountriesTable {...fakeProps} />
      </MemoryRouter>
    );
    const row = screen.getByTestId("row-CAN");
    fireEvent.click(row);
    expect(mockedNavigate).toHaveBeenCalledWith("/country/CAN");
  });


//CountryRow Component*********

test("if select country is called when a row item is clicked", () => {
    const mockSelectCountry = jest.fn()
    const fakeProps = {
      country: mockCountries[0],
      onSelectCountry: mockSelectCountry
    };
    render(
      <MemoryRouter>
        <CountryRow {...fakeProps}/>
      </MemoryRouter>
    );
    const row = screen.getByTestId("row-CAN");
    fireEvent.click(row);
    expect(mockSelectCountry).toHaveBeenCalledWith("CAN");
  });

  test("calls onSelectCountry when Enter key is pressed", () => { //making sure keyboard command works and not just click
    const mockSelect = jest.fn();
    const fakeProps = {
      country: mockCountries[0], 
      onSelectCountry: mockSelect,
    };
    render(
      <MemoryRouter>
        <CountryRow {...fakeProps} />
      </MemoryRouter>
    );
    const row = screen.getByTestId("row-CAN");
    row.focus(); 
    fireEvent.keyDown(row, { key: "Enter", code: "Enter", charCode: 13 });
    expect(mockSelect).toHaveBeenCalledWith("CAN");
  });

  test("renders the common and official names", () => {
    const mockSelect = jest.fn();
    const fakeProps = {
      country: mockCountries[0], 
      onSelectCountry: mockSelect,
    };
    render(
      <MemoryRouter>
        <CountryRow {...fakeProps} />
      </MemoryRouter>
    );
    const common = screen.getAllByText("Canada"); //each common name is in both title and the written text
    expect(common.length).toBeGreaterThan(0);
    const official = screen.getAllByText("Canadadada"); //each official name is in both title and the written text
    expect(official.length).toBeGreaterThan(0);
  });
  
  test("row is keyboard accessible", () => {
    const mockSelect = jest.fn();
    const fakeProps = {
      country: mockCountries[0], 
      onSelectCountry: mockSelect,
    };
    render(
      <MemoryRouter>
        <CountryRow {...fakeProps} />
      </MemoryRouter>
    );
    const row = screen.getByTestId("row-CAN");
    expect(row).toHaveAttribute("tabIndex", "0");
  });
  
  test("matches snapshot", () => {
    const { container } = render(
      <MemoryRouter>
        <CountryRow
          country={mockCountries[0]}
          onSelectCountry={() => {}}
        />
      </MemoryRouter>
    );
  
    expect(container).toMatchSnapshot();
  });

  
  //Pagination Component******
  test("renders correct number of page buttons", () => {
    render(
      <Pagination
      currentPage={1}
      setCurrentPage= {() => {}}
      amtCountriesPerPage={10}
      filteredCountries={mockCountries}
      />
    );
  
    const currPage = screen.getAllByText("1");
    expect(currPage.length).toBeGreaterThan(0);
    const totalPages = screen.getAllByText("1");
    expect(totalPages.length).toBeGreaterThan(0); //for different amounts, put more in the mock
  });

  test("calls setCurrent Page on Next button click", () => {
    const mockPageChange = jest.fn();
    render(
      <Pagination
      currentPage={1}
      setCurrentPage= {mockPageChange}
      amtCountriesPerPage={1}
      filteredCountries={mockCountries}
      />
    );

    fireEvent.click(screen.getByText(/next/i));
  expect(mockPageChange).toHaveBeenCalledWith(2);
});

test("calls setCurrent Page on Prev button click", () => {
    const mockPageChange = jest.fn();
    render(
      <Pagination
      currentPage={2}
      setCurrentPage= {mockPageChange}
      amtCountriesPerPage={1}
      filteredCountries={mockCountries}
      />
    );

    fireEvent.click(screen.getByText(/previous/i));
  expect(mockPageChange).toHaveBeenCalledWith(1);
});

test("Prev Button is disabled", () => {
    const mockPageChange = jest.fn();
    render(
      <Pagination
      currentPage={1}
      setCurrentPage= {mockPageChange}
      amtCountriesPerPage={1}
      filteredCountries={mockCountries}
      />
    );

    expect(screen.getByText(/previous/i)).toBeDisabled();

});

test("Next Button is disabled", () => {
    const mockPageChange = jest.fn();
    render(
      <Pagination
      currentPage={2}
      setCurrentPage= {mockPageChange}
      amtCountriesPerPage={1}
      filteredCountries={mockCountries}
      />
    );

    expect(screen.getByText(/next/i)).toBeDisabled();

});