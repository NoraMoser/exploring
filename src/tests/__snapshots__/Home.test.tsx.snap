// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <tr
    aria-label="view details for Canada"
    data-testid="row-CAN"
    style="cursor: pointer;"
    tabindex="0"
  >
    <td
      data-label="Common Name"
      title="Canada"
    >
      Canada
    </td>
    <td
      data-label="Official Name"
      title="Canadadada"
    >
      Canadadada
    </td>
  </tr>
</div>
`;
